Parser/Client/Browser.php

```diff
@@ -506,6 +518,138 @@ public static function isMobileOnlyBrowser(string $browser): bool
+            if ($name . ' Mobile' === $browserFromUserAgent['name']) {
+                $name  = $browserFromUserAgent['name'];
+                $short = $browserFromUserAgent['short_name'];
+            }
+
+            // If useragent detects another browser, but the family matches, we use the detected engine from useragent
+            if ($name !== $browserFromUserAgent['name']
+                && self::getBrowserFamily($name) === self::getBrowserFamily($browserFromUserAgent['name'])
+            ) {
+                $engine        = $browserFromUserAgent['engine'] ?? '';
+                $engineVersion = $browserFromUserAgent['engine_version'] ?? '';
+            }
+
+            if ($name === $browserFromUserAgent['name']) {
+                $engine        = $browserFromUserAgent['engine'] ?? '';
+                $engineVersion = $browserFromUserAgent['engine_version'] ?? '';
+
+                // In case the user agent reports a more detailed version, we try to use this instead
+                if (!empty($browserFromUserAgent['version'])
+                    && 0 === \strpos($browserFromUserAgent['version'], $version)
+                    && \version_compare($version, $browserFromUserAgent['version'], '<')
+                ) {
+                    $version = $browserFromUserAgent['version'];
+                }
+            }
+        } else {
+            $name          = $browserFromUserAgent['name'];
+            $version       = $browserFromUserAgent['version'];
+            $short         = $browserFromUserAgent['short_name'];
+            $engine        = $browserFromUserAgent['engine'];
+            $engineVersion = $browserFromUserAgent['engine_version'];
+        }
+
+        if (empty($name)) {
+            return [];
+        }
+
+        return [
+            'type'           => 'browser',
+            'name'           => $name,
+            'short_name'     => $short,
+            'version'        => $version,
+            'engine'         => $engine,
+            'engine_version' => $engineVersion,
+            'family'         => self::getBrowserFamily((string) $short),
+        ];
+    }
+
+    /**
+     * Returns the browser that can be safely detected from client hints
+     *
+     * @return array
+     */
+    protected function parseBrowserFromClientHints(): array
+    {
+        $name = $version = $short = '';
+
+        if ($this->clientHints instanceof ClientHints && $this->clientHints->getBrandList()) {
+            $brands = $this->clientHints->getBrandList();
+
+            foreach ($brands as $brand => $brandVersion) {
+                foreach (self::$clientHintMapping as $browserName => $clientHints) {
+                    foreach ($clientHints as $clientHint) {
+                        if (\strtolower($brand) === \strtolower($clientHint)) {
+                            $brand = $browserName;
+
+                            break 2;
+                        }
+                    }
+                }
```

Comment:
I would put this code in a separate method

```php
 move to AbstractClass
 create static var AbstractClass::$clientHintMapping
 protected function findNameByHintMapping(string $name): string 
    {
        foreach (self::$clientHintMapping as $mapName => $clientHints) {
            foreach ($clientHints as $clientHint) {
                if (\strtolower($name) === \strtolower($clientHint)) {
                    return $mapName;
                }
            }
        }
        return $name;
    }
}
```

```php
foreach ($brands as $brand => $brandVersion) {
    $brand = $this->findNameByHintMapping($brand);
}
```

Reply:
the same code can be used in OS Parser

Reply:
will change that.